audio/out/ao_coreaudio.c

```diff
@@ -78,7 +78,14 @@ static OSStatus render_cb_lpcm(void *ctx, AudioUnitRenderActionFlags *aflags,
 
     int64_t end = mp_time_ns();
     end += p->hw_latency_ns + ca_get_latency(ts) + ca_frames_to_ns(ao, frames);
-    ao_read_data(ao, planes, frames, end);
+    int samples = ao_read_data(ao, planes, frames, end);
+
+    if (samples == 0)
+        *aflags |= kAudioUnitRenderAction_OutputIsSilence;
+
+    for (int n = 0; n < buffer_list->mNumberBuffers; n++)
+        buffer_list->mBuffers[n].mDataByteSize = samples * ao->sstride;
```

Comment:
I'm not sure it is correct to send variable data rate to device. Is it handled correctly? Is it padded with silence?

I was looking at other issue, but I think we have underrun we should send silence.

Reply:
> I'm not sure it is correct to send variable data rate to device. Is it handled correctly? Is it padded with silence?

The only thing I know is that changing `mDataByteSize` fixes crackling. I don't know whether that's correct, because Apple docs are very sparse on this subject.

Not touching `mDataByteSize` and sending silence produces crackling.

I guess you can think of `mDataByteSize` as the value by which the write position to a large audio buffer is advanced. Before, we advanced by the requested amount of samples even if we couldn't write that many, producing gaps in the audio buffer which were never filled. With this change, we advance the write position by the exact number of samples we wrote. That's my theory though, I don't know the actual explanation.